src/Compilers/Core/Portable/DiagnosticAnalyzer/ArtifactProducer.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Text;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// The base type for artifact producers that can programmatically generate additional non-code files during a
+    /// compilation.  Artifact generators only run when a compiler is invoked with the <c>generatedartifactsout</c>
+    /// parameter.
+    /// </summary>
+    /// <remarks>
+    /// Normally, an artifact generator will not need to report diagnostics.  However, it may sometimes be necessary if
+    /// errors or other issues arise during generation.  If diagnostic reporting is needed, then the same mechanism are
+    /// available as with normal analyzers.  Specifically, <see cref="SupportedDiagnostics"/> should be overridden to
+    /// state which diagnostics may be produced, and the various context `ReportDiagnostic` calls should be used to
+    /// report them.
+    /// </remarks>
+    public abstract class ArtifactProducer : DiagnosticAnalyzer
```

Comment:
Is there a proposal document or thread somewhere I could look at for the design of this? At first glance, it seems odd that we'd have this subclass `DiagnosticAnalyzer`

Reply:
Going to hold off on further review until I have a picture of the actual proposal.

Reply:
1. What woudl be the best way to make a proposal here?
2. "At first glance, it seems odd that we'd have this subclass `DiagnosticAnalyzer`".  We discussed this in our design meetings.  It's unfortunate, but `DiagnosticAnalzyer` is really better named as `CompilerPlugin` :)  We already subclass it in several ways for this sort of approach.

Reply:
1. Perhaps send an email to the codeanalysis API review alias?
2. I was against us subclassing it for several of the other places that we subclass it there, and it felt we did it for the sake of expedience. I would prefer not to repeat if possible.

Reply:
> I was against us subclassing it for several of the other places that we subclass it there, and it felt we did it for the sake of expedience. I would prefer not to repeat if possible.

FYI: Not re-using the DiagnosticAnalyzer sub-type would essentially mean rewriting/cloning the entire analyzer driver as we want the artifact producer to be able to register for all regular analysis callbacks that analyzers can register for analysis. This is extremely large amount of work, not to mention the large maintenance cost that it will incur in future.

Reply:
Shoudl we introduce a new base class *above* DiagnosticAnalyser and update things to use that?  i.e.  `public abstract class CompilerPlugin`?  We can hten make ArtifactProducer subclass that new `CompilerPlugin`? 

This could be done in two PRs.  The first introduces the subclass, then this PR updates to use that?

Reply:
> Shoudl we introduce a new base class above DiagnosticAnalyser and update things to use that?

I like the sound of that proposal.